### PR TITLE
[typed-store] implement the prev() on iterator

### PR DESCRIPTION
**Pull request by akichidis**
Introduces the method `prev()` on the iterator used by the `DBMap` so we can retrieve our results in reverse order. That's particularly useful in situations where you want to `skip_to_last` element and start retrieving the elements backwards.

### CODE REVIEW

**Review by huitseeker — COMMENTED** [review]
I see no objection with this but I note that the idiomatic way to express this is to [implement a `DoubleEndedIterator`](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html) rather than add a method on the `Iter` implementation which would be difficult to discover.
What do you think about this approach?

**Comment by akichidis** [comment]
> I see no objection with this but I note that the idiomatic way to express this is to [implement a `DoubleEndedIterator`](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html) rather than add a method on the `Iter` implementation which would be difficult to discover. What do you think about this approach?

Ah nice, thanks @huitseeker for pointing me on the DoubleEndedIterator, wasn't aware of it, makes total sense! I am updating the PR now

**Review by huitseeker — COMMENTED** [review]
I've added a test to the end of this PR, illustrating why I think the behavior of the `DoubleEndedIterator` doesn't match its specification - the implementation iterates in the reverse direction, but is supposed to remove from the end of the iterator.
This may require modifying the `Iter` struct to keep two pointers to iterators with distinct [`IteratorMode`](https://docs.rs/rocksdb/latest/rocksdb/enum.IteratorMode.html) values.

(should you not want to get into it at this time, tabling this and going back to an implementation of `prev` is fine too)

File: crates/typed-store/src/rocks/tests.rs [hunk 1]
```diff
@@ -233,6 +233,26 @@ fn test_iter() {
...
+
+    db.insert(&1, &"1".to_string()).expect("Failed to insert");
+    db.insert(&2, &"2".to_string()).expect("Failed to insert");
+    db.insert(&3, &"3".to_string()).expect("Failed to insert");
+
+    let mut iter = db.iter().skip_to_last().reverse();
+    assert_eq!(Some((3, "3".to_string())), iter.next());
+    assert_eq!(Some((2, "2".to_string())), iter.next());
+    assert_eq!(Some((1, "1".to_string())), iter.next());
+    assert_eq!(None, iter.next());
```
Comment (akichidis) [inline crates/typed-store/src/rocks/tests.rs:248]: I believe this is now more _natural_ when it comes to iterate the elements in reverse order. Thoughts @huitseeker ?

File: crates/typed-store/src/rocks/iter.rs [hunk 2]
```diff
@@ -68,4 +75,35 @@ impl<'a, K: Serialize, V> Iter<'a, K, V> {
...
+        self.direction = Direction::Reverse;
+        RevIter::new(self)
+    }
+}
+
+/// An iterator with a reverted direction to the original. The `RevIter`
+/// is hosting an iteration which is consuming in the opposing direction.
+/// It's not possible to do further manipulation (ex re-reverse) to the
+/// iterator.
+pub struct RevIter<'a, K, V> {
```
Comment (akichidis) [inline crates/typed-store/src/rocks/iter.rs:92]: Introduced a separate struct so can limit the user from doing any complex combinations of seeks that could confuse the iterator. Now there is only one option once you reverse...you are calling `next` to traverse the results backwards

File: crates/typed-store/src/rocks/tests.rs (hunk 1 above)
Comment (huitseeker) [reply crates/typed-store/src/rocks/tests.rs:248]: 👍  Yep. Perhaps this departs slightly from the rust idiom, but this is definitely correct, which is what I think we care about.

**Review by huitseeker — APPROVED** [review]
Thanks a bunch @akichidis 🙏

File: crates/typed-store/src/rocks/tests.rs (hunk 1 above)
Comment (akichidis) [reply crates/typed-store/src/rocks/tests.rs:248]: Ah yeah it's not very Rusty , but tried to go for correctness here as you said

**Pull request merged**